File: appointments-service/src/main/kotlin/it/nucleo/appointments/infrastructure/kafka/NotificationEventsPublisher.kt
package it.nucleo.appointments.infrastructure.kafka

import java.time.Instant
import java.util.Properties
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import org.apache.kafka.clients.producer.KafkaProducer
import org.apache.kafka.clients.producer.ProducerConfig
import org.apache.kafka.clients.producer.ProducerRecord
import org.apache.kafka.common.serialization.StringSerializer
import org.slf4j.LoggerFactory

@Serializable
private data class NotificationEvent(
    val receiver: String,
    val title: String,
    val content: String? = null,
    val sourceService: String,
    val occurredAt: String
)

class NotificationEventsPublisher(
    private val bootstrapServers: String,
    private val clientId: String,
    private val notificationsTopic: String
) {
    private val logger = LoggerFactory.getLogger(NotificationEventsPublisher::class.java)
    private val json = Json { encodeDefaults = true }
    private var producer: KafkaProducer<String, String>? = null

    fun isEnabled(): Boolean {
        return bootstrapServers.isNotBlank() && notificationsTopic.isNotBlank()
    }

    fun publish(receiver: String, title: String, content: String? = null) {
        if (!isEnabled() || receiver.isBlank() || title.isBlank()) {
            return
        }

        try {
            val event =
                NotificationEvent(
                    receiver = receiver,
                    title = title,
                    content = content,
                    sourceService = clientId,
                    occurredAt = Instant.now().toString()
                )
            val payload = json.encodeToString(event)

            getOrCreateProducer().send(
                ProducerRecord(notificationsTopic, receiver, payload)
            )
        } catch (error: Exception) {
            logger.error("Failed to publish notification event", error)
        }
    }

    fun close() {
        producer?.close()
        producer = null
    }

    private fun getOrCreateProducer(): KafkaProducer<String, String> {
        val currentProducer = producer
        if (currentProducer != null) {
            return currentProducer
        }

        val properties = Properties()
        properties[ProducerConfig.BOOTSTRAP_SERVERS_CONFIG] = bootstrapServers
        properties[ProducerConfig.CLIENT_ID_CONFIG] = "$clientId-notifications"
        properties[ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG] = StringSerializer::class.java
        properties[ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG] = StringSerializer::class.java

        val createdProducer = KafkaProducer<String, String>(properties)
        producer = createdProducer
        return createdProducer
    }
}

File: users-service/src/infrastructure/kafka/notification-events.publisher.ts
import { Kafka, type Producer } from 'kafkajs';

interface NotificationEvent {
  receiver: string;
  title: string;
  content: string | null;
  sourceService: string;
  occurredAt: string;
}

export class NotificationEventsPublisher {
  private readonly bootstrapServers: string;
  private readonly clientId: string;
  private readonly notificationsTopic: string;
  private producer: Producer | null;
  private isConnected: boolean;

  constructor() {
    this.bootstrapServers = process.env.KAFKA_BOOTSTRAP_SERVERS ?? '';
    this.clientId = process.env.KAFKA_CLIENT_ID ?? 'users-service';
    this.notificationsTopic = process.env.KAFKA_TOPIC_NOTIFICATIONS ?? '';
    this.producer = null;
    this.isConnected = false;
  }

  async publish(payload: {
    receiver: string;
    title: string;
    content?: string | null;
  }): Promise<void> {
    if (!this.isEnabled() || !payload.receiver.trim() || !payload.title.trim()) {
      return;
    }

    await this.ensureConnected();

    const event: NotificationEvent = {
      receiver: payload.receiver.trim(),
      title: payload.title.trim(),
      content: payload.content ?? null,
      sourceService: this.clientId,
      occurredAt: new Date().toISOString(),
    };

    await this.producer?.send({
      topic: this.notificationsTopic,
      messages: [
        {
          key: event.receiver,
          value: JSON.stringify(event),
        },
      ],
    });
  }

  async disconnect(): Promise<void> {
    if (!this.producer || !this.isConnected) {
      return;
    }

    await this.producer.disconnect();
    this.isConnected = false;
  }

  private isEnabled(): boolean {
    return Boolean(this.bootstrapServers.trim()) && Boolean(this.notificationsTopic.trim());
  }

  private async ensureConnected(): Promise<void> {
    if (this.isConnected) {
      return;
    }

    const kafka = new Kafka({
      clientId: `${this.clientId}-notifications`,
      brokers: this.bootstrapServers.split(',').map(function (broker) {
        return broker.trim();
      }),
    });

    this.producer = kafka.producer();
    await this.producer.connect();
    this.isConnected = true;
  }
}
